qiskit/circuit/library/data_preparation/state_preparation.py

```diff
@@ -0,0 +1,505 @@
+from qiskit.circuit.library.standard_gates.s import SGate, SdgGate
+from qiskit.circuit.library.standard_gates.ry import RYGate
+from qiskit.circuit.library.standard_gates.rz import RZGate
+from qiskit.circuit.exceptions import CircuitError
+from qiskit.quantum_info import Statevector
+
+_EPS = 1e-10  # global variable used to chop very small numbers to zero
+
+
+class StatePreparation(Gate):
+    """Complex amplitude state preparation.
+
+    Class that implements the (complex amplitude) state preparation of some
+    flexible collection of qubit registers.
+    """
+
+    def __init__(
+        self,
+        params: Union[str, list, int, Statevector],
+        num_qubits: Optional[int] = None,
+        inverse: bool = False,
+        label: Optional[str] = "State Preparation",
+    ):
+        r"""
+        Args:
+            params:
+                * Statevector: Statevector to initialize to.
+                * list: vector of complex amplitudes to initialize to.
+                * string: labels of basis states of the Pauli eigenstates Z, X, Y. See
+                  :meth:`.Statevector.from_label`.
+                  Notice the order of the labels is reversed with respect to the qubit index to
+                  be applied to. Example label '01' initializes the qubit zero to :math:`|1\rangle`
+                  and the qubit one to :math:`|0\rangle`.
+                * int: an integer that is used as a bitmap indicating which qubits to initialize
+                  to :math:`|1\rangle`. Example: setting params to 5 would initialize qubit 0 and qubit 2
+                  to :math:`|1\rangle` and qubit 1 to :math:`|0\rangle`.
+            num_qubits: This parameter is only used if params is an int. Indicates the total
+                number of qubits in the `initialize` call. Example: `initialize` covers 5 qubits
+                and params is 3. This allows qubits 0 and 1 to be initialized to :math:`|1\rangle`
+                and the remaining 3 qubits to be initialized to :math:`|0\rangle`.
+            inverse: if True, the inverse state is constructed.
+            label: An optional label for the gate [Default: State Preparation]
+
+        Raises:
+            QiskitError: ``num_qubits`` parameter used when ``params`` is not an integer
+
+        When a Statevector argument is passed the state is prepared using a recursive
+        initialization algorithm, including optimizations, from [1], as well
+        as some additional optimizations including removing zero rotations and double cnots.
+
+        **References:**
+        [1] Shende, Bullock, Markov. Synthesis of Quantum Logic Circuits (2004)
+        [`https://arxiv.org/abs/quant-ph/0406176v5`]
+
+        """
+        self._params_arg = params
+        self._inverse = inverse
+        self._name = "state_preparation_dg" if self._inverse else "state_preparation"
+        self._label = f"{label} Dg" if self._inverse else label
```

Comment:
This will cause `StatePreparation().inverse().inverse()` to have a label `"StatePreparation Dg Dg"` if you care about that sort of thing (I probably wouldn't bother).